File: src/simulator/Robot.java
package simulator;

import math.Vec2;
import math.Vec3;
import processing.core.PApplet;

import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

public class Robot {
    public static final double MAX_LINEAR_ACCELERATION = 0.5;
    public static final double MAX_ANGULAR_ACCELERATION = 0.5;
    public static final double LINEAR_VELOCITY_ERROR_LIMIT = 0.1;
    public static final double ANGULAR_VELOCITY_ERROR_LIMIT = 0.1;

    // Graphics
    private final PApplet applet;

    // Main frame
    private final double robotLength;
    private Vec3 truePose;
    private boolean isRunning;

    // Multi-thread access
    private final Vec2 goalControl = Vec2.zero();
    private final Vec2 currentControl = Vec2.zero();

    // Sensors
    final Laser laser;

    Robot(PApplet applet, double robotLength, Vec3 truePose, boolean isRunning) {
        this.applet = applet;
        this.robotLength = robotLength;
        this.truePose = Vec3.of(truePose);
        this.isRunning = isRunning;
        this.laser = new Laser(applet);
    }

    boolean isRunning() {
        return isRunning;
    }

    void setRunning(boolean running) {
        isRunning = running;
    }

    private static Vec3 getChangeInPose(Vec3 pose, Vec2 control) {
        Vec3 changeInPose = Vec3.zero();
        changeInPose.x = control.x * Math.cos(pose.z);
        changeInPose.y = control.x * Math.sin(pose.z);
        changeInPose.z = control.y;
        return changeInPose;
    }

    void updatePose(double dt) {
        Vec2 tmpControl = Vec2.zero();
        synchronized (currentControl) {
            synchronized (goalControl) {
                // Only allow so much acceleration per timestep
                Vec2 control_diff = goalControl.minus(currentControl);
                if (Math.abs(control_diff.x) > Robot.MAX_LINEAR_ACCELERATION * dt) {
                    control_diff.x = Math.signum(control_diff.x) * Robot.MAX_LINEAR_ACCELERATION * dt;
                }
                if (Math.abs(control_diff.y) > Robot.MAX_ANGULAR_ACCELERATION * dt) {
                    control_diff.y = Math.signum(control_diff.y) * Robot.MAX_ANGULAR_ACCELERATION * dt;
                }
                currentControl.plusInPlace(control_diff);
                tmpControl = currentControl;
            }
        }

        // Only apply noise if we're trying to move
        if (tmpControl.sqauredNorm() != 0.0) {
            tmpControl.x *= (1.0 + ThreadLocalRandom.current().nextDouble(-LINEAR_VELOCITY_ERROR_LIMIT, LINEAR_VELOCITY_ERROR_LIMIT));
            tmpControl.y *= (1.0 + ThreadLocalRandom.current().nextDouble(-ANGULAR_VELOCITY_ERROR_LIMIT, ANGULAR_VELOCITY_ERROR_LIMIT));
        }

        // Run the dynamics via RK4
        Vec3 k1, k2, k3, k4;
        Vec3 x2, x3, x4;
        k1 = getChangeInPose(truePose, tmpControl);
        x2 = truePose.plus(k1.scale(0.5f * dt));
        k2 = getChangeInPose(x2, tmpControl);
        x3 = truePose.plus(k2.scale(0.5f * dt));
        k3 = getChangeInPose(x3, tmpControl);
        x4 = truePose.plus(k3.scale(dt));
        k4 = getChangeInPose(x4, tmpControl);

        Vec3 dtruePose = Vec3.zero();
        dtruePose
                .plusInPlace(k1)
                .plusInPlace(k2.scale(2))
                .plusInPlace(k3.scale(2))
                .plusInPlace(k4)
                .scaleInPlace(dt / 6.0);
        truePose.plusInPlace(dtruePose);
    }

    void updateSense(List<Landmark> landmarks) {
        // Move the center of the scanner back from the center of the robot
        Vec2 truePosition = Vec2.of(truePose.x, truePose.y);
        Vec2 laserCenter = truePosition.minus(Vec2.of(Math.cos(truePose.z), Math.sin(truePose.z)).scaleInPlace(0.5 * robotLength));
        laser.updateLaserScan(laserCenter, truePose.z, landmarks);
    }

    boolean isCrashing(Landmark landmark) {
        return landmark.shortestDistanceFrom(Vec2.of(truePose.x, truePose.y)) < robotLength;
    }

    Vec2 getCurrentControl() {
        Vec2 answer = Vec2.zero();
        synchronized (currentControl) {
            answer.set(currentControl);
        }
        return answer;
    }

    void applyControl(Vec2 control) {
        synchronized (goalControl) {
            goalControl.set(control);
        }
    }

    void draw() {
        Vec2 position = Vec2.of(truePose.x, truePose.y).scaleInPlace(Simulator.SCALE).plusInPlace(Vec2.of(Simulator.WIDTH / 2.0, Simulator.HEIGHT / 2.0));
        Vec2 laserEnd = position.minus(Vec2.of(Math.cos(truePose.z), Math.sin(truePose.z)).scaleInPlace(0.5 * robotLength * Simulator.SCALE));
        Vec2 otherEnd = position.plus(Vec2.of(Math.cos(truePose.z), Math.sin(truePose.z)).scaleInPlace(0.5 * robotLength * Simulator.SCALE));

        // Draw lasers
        laser.draw(laserEnd, truePose.z);

        // Draw robot body
        applet.stroke(1);
        applet.line((float) laserEnd.x, (float) laserEnd.y, (float) otherEnd.x, (float) otherEnd.y);
    }
}

File: src/simulator/Robot.kt
package simulator

import extensions.*
import org.ejml.data.FMatrix2
import org.ejml.data.FMatrix3
import processing.core.PApplet
import java.util.concurrent.ThreadLocalRandom
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sign
import kotlin.math.sin

class Robot internal constructor(private val applet: PApplet, val length: Float, truePose: FMatrix3, var isRunning: Boolean) {
    companion object {
        const val MAX_LINEAR_ACCELERATION = 200000f
        const val MAX_ANGULAR_ACCELERATION = 5f
        const val LINEAR_VELOCITY_ERROR_LIMIT = 2f
        const val ANGULAR_VELOCITY_ERROR_LIMIT = 0.1f
        private fun getChangeInPose(pose: FMatrix3, control: FMatrix2): FMatrix3 {
            val changeInPose = FMatrix3()
            changeInPose.a1 = control.a1 * cos(pose.a3)
            changeInPose.a2 = control.a1 * sin(pose.a3)
            changeInPose.a3 = control.a2
            return changeInPose
        }

        private fun getPositionFromPose(pose: FMatrix3): FMatrix2 {
            return FMatrix2(pose.a1, pose.a2)
        }
    }

    // Multi-thread access
    private val truePose = FMatrix3(truePose)
    private val goalControl = FMatrix2()
    private val currentControl = FMatrix2()

    // Sensors
    val laserSensor: LaserSensor = LaserSensor(applet)

    fun updatePose(dt: Float) {
        var noisyControl = FMatrix2()
        synchronized(currentControl) {
            synchronized(goalControl) {

                // Clamp acceleration
                val controlDiff = goalControl - currentControl
                if (abs(controlDiff.a1) > MAX_LINEAR_ACCELERATION * dt) {
                    controlDiff.a1 = sign(controlDiff.a1) * MAX_LINEAR_ACCELERATION * dt
                }
                if (abs(controlDiff.a2) > MAX_ANGULAR_ACCELERATION * dt) {
                    controlDiff.a2 = sign(controlDiff.a2) * MAX_ANGULAR_ACCELERATION * dt
                }
                currentControl += controlDiff
                noisyControl = FMatrix2(currentControl)
            }
        }

        // Only apply noise if we're trying to move
        if (noisyControl.squaredNorm() != 0f) {
            noisyControl.a1 *= 1f + ThreadLocalRandom.current().nextDouble(-LINEAR_VELOCITY_ERROR_LIMIT.toDouble(), LINEAR_VELOCITY_ERROR_LIMIT.toDouble()).toFloat()
            noisyControl.a2 *= 1f + ThreadLocalRandom.current().nextDouble(-ANGULAR_VELOCITY_ERROR_LIMIT.toDouble(), ANGULAR_VELOCITY_ERROR_LIMIT.toDouble()).toFloat()
        }

        // Run the dynamics via RK4
        synchronized(truePose) {
            val k1 = getChangeInPose(truePose, noisyControl)
            val x2 = truePose + k1 * (0.5f * dt)
            val k2 = getChangeInPose(x2, noisyControl)
            val x3 = truePose + k2 * (0.5f * dt)
            val k3 = getChangeInPose(x3, noisyControl)
            val x4 = truePose + k3 * dt
            val k4 = getChangeInPose(x4, noisyControl)
            val avgChangeInPose = FMatrix3()
            avgChangeInPose += k1
            avgChangeInPose += k2 * 2f
            avgChangeInPose += k3 * 2f
            avgChangeInPose += k4
            avgChangeInPose *= dt / 6f
            truePose += avgChangeInPose
        }
    }

    fun updateSense(obstacles: List<Obstacle>) {
        // Move the center of the scanner back from the center of the robot
        val truePosition: FMatrix2
        val orientation: Float
        synchronized(truePose) {
            truePosition = getPositionFromPose(truePose)
            orientation = truePose.a3
        }
        val centerToHead = FMatrix2(cos(orientation), sin(orientation))
        centerToHead *= 0.5f * length
        val laserSource = truePosition - centerToHead
        laserSensor.updateLaserScan(laserSource, orientation, obstacles)
    }

    fun isCrashing(obstacle: Obstacle): Boolean {
        synchronized(truePose) {
            return obstacle.shortestDistanceFrom(getPositionFromPose(truePose)) < length / 2
        }
    }

    /* User callable */
    fun getTruePose(): FMatrix3 {
        val answer = FMatrix3()
        synchronized(truePose) { answer.set(truePose) }
        return answer
    }

    fun getCurrentControl(): FMatrix2 {
        val answer = FMatrix2()
        synchronized(currentControl) { answer.set(currentControl) }
        return answer
    }

    fun applyControl(control: FMatrix2) {
        synchronized(goalControl) { goalControl.set(control) }
    }

    fun draw() {
        val truePoseCopy = getTruePose()
        val center = getPositionFromPose(truePoseCopy)
        val centerToHeadUnit = FMatrix2(cos(truePoseCopy.a3), sin(truePoseCopy.a3))
        centerToHeadUnit *= 0.5f * length
        val head = center + centerToHeadUnit
        val tail = center - centerToHeadUnit

        // Draw robot body
        applet.noFill()
        applet.stroke(1)
        applet.line(center.a1, 0f, center.a2, head.a1, 0f, head.a2)
        applet.circleXZ(center.a1, center.a2, length * 0.5f)

        // Draw lasers
        laserSensor.draw(tail)
    }

}